### PR TITLE
Migrate existing DB system to use Doctrine ORM

### PR DESCRIPTION
**Pull request by MusikAnimal**
Migrate existing DB system to use Doctrine ORM

This moves App\CreationLog to App\Entity\GeneratedBook and
App\Repository\GeneratedBookRepository.

Bump composer packages (resolving the merge conflict)

Silence indirect deprecations in tests so that the CI build won't fail.

Bug: T266190

### CODE REVIEW

**Comment by MusikAnimal** [comment]
WIP. Maybe went a little too far with the Entity abstraction, but it felt "correct" if we wanted to appease the Symfony gods.

File: config/services.yaml [hunk 1]
```diff
@@ -29,7 +29,7 @@ services:
         resource: '../src/Controller/'
         tags: ['controller.service_arguments']
 
-    App\CreationLog:
+    App\Controller\ExportController:
         arguments:
             $enableStats: '%env(bool:APP_ENABLE_STATS)%'
```
Comment (samwilson) [inline config/services.yaml:34]: I wonder if there's any point in being able to disable stats collection?  It seems like we could probably just drop this env var (could be a future patch of course, or even one before this one).

File: migrations/Version20201216012133.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+final class Version20201216012133 extends AbstractMigration
+{
+    public function getDescription() : string
+    {
+        return '';
+    }
+
+    public function up(Schema $schema) : void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
```
Comment (samwilson) [inline migrations/Version20201216012133.php:22]: Remove auto comments?

File: src/Entity/CreationLog.php [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+ * @ORM\Entity(repositoryClass="App\Repository\CreationLogRepository")
+ * @ORM\Table(
+ *     name="books_generated",
+ *     indexes={
+ *         @ORM\Index(name="time", columns={"time"}),
+ *         @ORM\Index(name="lang", columns={"lang"})
+ *     }
+ * )
+ */
+class CreationLog {
```
Comment (samwilson) [inline src/Entity/CreationLog.php:21]: Entities are usually singular names of the records inside them. Maybe `GeneratedBook` (to be similar to the table name)?

**Review by samwilson — COMMENTED** [review]
I know you're not looking for reviews yet, but I just saw a couple of things.

File: config/services.yaml (hunk 1 above)
Comment (MusikAnimal) [reply config/services.yaml:34]: Funny you ask; tools-db has gone down several times lately. Did you or someone disable this, or is there some other magic that makes it not error out? I don't remember seeing any errors. Anyway, that's what I figured this env var was for.

File: src/Entity/CreationLog.php (hunk 3 above)
Comment (MusikAnimal) [reply src/Entity/CreationLog.php:21]: :+1: I wanted to change it but wasn't sure if it'd be too different than what people are used to. `GeneratedBook` sounds good!

**Comment by MusikAnimal** [comment]
Tests pass on my local, though with a LOT of seemingly unrelated warnings.

**Comment by MusikAnimal** [comment]
I don't know what to do about the deprecation notice that's causing CI to fail: https://github.com/wsexport/tool/pull/291/checks?check_run_id=1574104284. That's in code we didn't write. dama/doctrine-test-bundle is up-to-date, the latest release came just six days ago. Should we [disable the deprecations helper](https://symfony.com/doc/current/components/phpunit_bridge.html#disabling-the-deprecation-helper) for now?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -53,6 +53,7 @@ jobs:
     - name: Install WSExport
       run: |
         composer install
+        ./bin/console doctrine:migrations:migrate --env=test --no-interaction
```
Comment (samwilson) [inline .github/workflows/ci.yml:56]: APP_ENV is already being set above in the `env` bit, it shouldn't be needed here.

File: src/Entity/GeneratedBook.php [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	 */
+	private $format;
+
+	/**
+	 * GeneratedBook constructor.
+	 * @param Book $book
+	 * @param string $format
+	 */
+	public function __construct( Book $book, string $format ) {
+		$this->time = new DateTime( 'now', new DateTimeZone( 'UTC' ) );
```
Comment (samwilson) [inline src/Entity/GeneratedBook.php:63]: Why not use setters for all of these? It reads a bit oddly that only Title does so.

File: src/Entity/GeneratedBook.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+	 */
+	public function getTitle(): string {
+		return $this->title;
+	}
+
+	/**
+	 * @param string $title
+	 */
+	public function setTitle( string $title ): void {
+		$this->title = htmlspecialchars_decode( $title );
```
Comment (samwilson) [inline src/Entity/GeneratedBook.php:115]: I think `htmlspecialchars_decode` doesn't belong here, unless it's made obvious that the provided $title is expected to be encoded (e.g. by a different method name). The entity should be usable from wherever.

Not that there are many books like "The Life of %20" but who knows... :)

**Review by samwilson — COMMENTED** [review]
Looks great!

How do we handle migrating the first time? After the `doctrine_migration_versions` table has been created, I guess we'll need to manually add the first migration (in prod anyway; in staging or dev we can just nuke the existing table probably). Or we could change the first migration to do "CREATE TABLE IF NOT EXISTS", then it wouldn't matter if it runs or not.

File: src/Entity/GeneratedBook.php (hunk 6 above)
Comment (MusikAnimal) [reply src/Entity/GeneratedBook.php:115]: This came from https://github.com/wsexport/tool/blob/151eec557a9dcc5198a4a0652a12db5edea188bb/src/CreationLog.php#L60

I didn't want to break anything by removing that. From https://phabricator.wikimedia.org/T270495 it sounds like the titles are sometimes encoded?

Comment (samwilson) [reply src/Entity/GeneratedBook.php:115]: Ah right! Makes sense. Let's keep it.

**Review by samwilson — APPROVED** [review]

File: src/Entity/CreationLog.php (hunk 3 above)
Comment (Fuerdai-Vandal) [reply src/Entity/CreationLog.php]: FUERDAI IS HERE FROM WIKIPEDIA AND WILL VANDALISE GITHUB ON WHEELS

File: src/Controller/StatisticsController.php [hunk 7]
```diff
@@ -20,14 +23,19 @@ private function normalizeFormat( string $format ): string {
...
 	 * @Route("/statistics", name="statistics")
 	 * @Route("/stat.php")
+	 * @param Request $request
+	 * @param EntityManager $entityManager
+	 * @return \Symfony\Component\HttpFoundation\Response
 	 */
-	public function index( Request $request, CreationLog $creationLog ) {
+	public function index( Request $request, EntityManagerInterface $entityManager ) {
+		/** @var GeneratedBookRepository $generatedBookRepo */
+		$generatedBookRepo = $entityManager->getRepository( GeneratedBook::class );
```
[the remark this answers is not in the document]
Comment (Fuerdai-Vandal) [reply src/Controller/StatisticsController.php]: I AM THE FUERDAI VANDAL ON WHEELS

File: .github/workflows/ci.yml (hunk 4 above)
Comment (Fuerdai-Vandal) [reply .github/workflows/ci.yml]: FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI

File: migrations/Version20201216012133.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+	public function getDescription() : string {
+		return 'Create the books_generated table';
+	}
+
+	/**
+	 * Creates the table.
+	 * @param Schema $schema
+	 */
+	public function up( Schema $schema ) : void {
+		$this->addSql( 'CREATE TABLE books_generated (time DATETIME NOT NULL, lang VARCHAR(10) NOT NULL, title VARCHAR(255) NOT NULL, format VARCHAR(10) NOT NULL, INDEX time (time), INDEX lang (lang)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_unicode_ci` ENGINE = InnoDB' );
```
[the remark this answers is not in the document]
Comment (Fuerdai-Vandal) [reply migrations/Version20201216012133.php]: WILLY ON WHEELS!!!

File: migrations/Version20201216031438.php [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+	public function getDescription() : string {
+		return 'Add primary key column to books_generated';
+	}
+
+	/**
+	 * Add the id column.
+	 * @param Schema $schema
+	 */
+	public function up( Schema $schema ) : void {
+		$this->addSql( 'ALTER TABLE books_generated ADD id INT AUTO_INCREMENT NOT NULL, ADD PRIMARY KEY (id)' );
```
[the remark this answers is not in the document]
Comment (Fuerdai-Vandal) [reply migrations/Version20201216031438.php]: FALUN GONG IS NOW CALLED FROG GONG!!!
DRINKING BLEACH CURES CORONAVIRUS

File: src/Entity/GeneratedBook.php (hunk 5 above)
Comment (Fuerdai-Vandal) [reply src/Entity/GeneratedBook.php]: MUSIKANIMAL IS A WIKIPEDOPHILE

File: config/services.yaml (hunk 1 above)
Comment (Fuerdai-Vandal) [reply config/services.yaml:41]: FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI FUERDAI

File: migrations/Version20201216012133.php (hunk 2 above)
Comment (Fuerdai-Vandal) [reply migrations/Version20201216012133.php]: MusikAnimal is a Wikipedophile.  I hate Wikipedia Administrators who keep blocking me.  I have every right to vandalise Wikipedia and you have no right to stop me.  ON WHEELS

**Pull request merged**